{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "    Author: Iris Shakya\n",
    "    course: Big Data Technology\n",
    "    Project: 2, tf-idf w/ similarity search( mapreduce + pyspark)\n",
    "    instructor: Prof. Lei Xei\n",
    "    References: https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.zipWithIndex.html\n",
    "                https://towardsdatascience.com/tf-idf-calculation-using-map-reduce-algorithm-in-pyspark-e89b5758e64c\n",
    "\"\"\"\n",
    "import math\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22/11/17 14:09:29 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n"
     ]
    }
   ],
   "source": [
    "from pyspark import SparkConf, SparkContext\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pyspark import SparkContext\n",
    "from pyspark.sql import SparkSession\n",
    "import gc\n",
    "from pyspark.sql.functions import monotonically_increasing_id\n",
    "\n",
    "sc = SparkContext(master = 'local')\n",
    "spark = SparkSession.builder.appName('pySpark word-count').config('spark.some.config.option', 'some-value')             .getOrCreate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = spark.sparkContext.textFile(\"/Users/iris/project2_demo.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "import regex as re\n",
    "regex_st = r'\\d{8}'\n",
    "\n",
    "# doc = data.map(lambda x: x.split(regex_st))\n",
    "doc = data.flatMap(lambda x: x.split())\n",
    "doc1 = doc.map(lambda x: (x.split(regex_st)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_para = data.map(lambda x: tuple(x.split('\\n')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(0,\n",
       "  ('10022814',\n",
       "   'growth',\n",
       "   'inhibition',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'grb2',\n",
       "   'downregulation',\n",
       "   'correlate',\n",
       "   'inactivation',\n",
       "   'mitogen-activated',\n",
       "   'protein',\n",
       "   'kinase',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'cell',\n",
       "   'increased',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'growth',\n",
       "   'associate',\n",
       "   'increase',\n",
       "   'expression',\n",
       "   'gene_epidermal_growth_factor_gene',\n",
       "   'receptor',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'receptor',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'rtks',\n",
       "   'upon',\n",
       "   'activation',\n",
       "   'rtks',\n",
       "   'transmit',\n",
       "   'oncogenic',\n",
       "   'signal',\n",
       "   'binding',\n",
       "   'growth',\n",
       "   'factor',\n",
       "   'receptor',\n",
       "   'bind',\n",
       "   'protein-2',\n",
       "   'grb2',\n",
       "   'turn',\n",
       "   'bind',\n",
       "   'sos',\n",
       "   'activate',\n",
       "   'ras-raf-mek-mitogen-activated',\n",
       "   'protein',\n",
       "   'map',\n",
       "   'kinase',\n",
       "   'pathway',\n",
       "   'grb2',\n",
       "   'important',\n",
       "   'transformation',\n",
       "   'fibroblast',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'whether',\n",
       "   'grb2',\n",
       "   'important',\n",
       "   'proliferation',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'express',\n",
       "   'rtks',\n",
       "   'unclear',\n",
       "   'use',\n",
       "   'liposome',\n",
       "   'deliver',\n",
       "   'nuclease-resistant',\n",
       "   'antisense',\n",
       "   'oligodeoxynucleotide',\n",
       "   'oligos',\n",
       "   'specific',\n",
       "   'grb2',\n",
       "   'mrna',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'grb2',\n",
       "   'protein',\n",
       "   'downregulation',\n",
       "   'inhibit',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'growth',\n",
       "   'degree',\n",
       "   'growth',\n",
       "   'inhibition',\n",
       "   'dependent',\n",
       "   'upon',\n",
       "   'activation',\n",
       "   'and-or',\n",
       "   'endogenou',\n",
       "   'level',\n",
       "   'rtks',\n",
       "   'grb2',\n",
       "   'inhibition',\n",
       "   'lead',\n",
       "   'map',\n",
       "   'kinase',\n",
       "   'inactivation',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'suggest',\n",
       "   'different',\n",
       "   'pathway',\n",
       "   'use',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'regulate',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'growth')),\n",
       " (1,\n",
       "  ('10022831',\n",
       "   'role',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'integrin',\n",
       "   'activation',\n",
       "   'vascular',\n",
       "   'endothelial',\n",
       "   'growth',\n",
       "   'factor',\n",
       "   'receptor-2',\n",
       "   'interaction',\n",
       "   'between',\n",
       "   'integrin',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'extracellular',\n",
       "   'matrix',\n",
       "   'crucial',\n",
       "   'endothelial',\n",
       "   'cell',\n",
       "   'sprout',\n",
       "   'capillary',\n",
       "   'angiogenesis',\n",
       "   'furthermore',\n",
       "   'integrin-mediated',\n",
       "   'outside-in',\n",
       "   'signal',\n",
       "   'co-operate',\n",
       "   'growth',\n",
       "   'factor',\n",
       "   'receptor',\n",
       "   'promote',\n",
       "   'cell',\n",
       "   'proliferation',\n",
       "   'motility',\n",
       "   'determine',\n",
       "   'potential',\n",
       "   'regulation',\n",
       "   'angiogenic',\n",
       "   'inducer',\n",
       "   'receptor',\n",
       "   'integrin',\n",
       "   'system',\n",
       "   'investigate',\n",
       "   'interaction',\n",
       "   'between',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'integrin',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'vascular',\n",
       "   'endothelial',\n",
       "   'growth',\n",
       "   'factor',\n",
       "   'receptor-2',\n",
       "   'vegfr-2',\n",
       "   'human',\n",
       "   'endothelial',\n",
       "   'cell',\n",
       "   'report',\n",
       "   'tyrosine-phosphorylated',\n",
       "   'vegfr-2',\n",
       "   'co-immunoprecipitated',\n",
       "   'beta3',\n",
       "   'integrin',\n",
       "   'subunit',\n",
       "   'beta1',\n",
       "   'beta5',\n",
       "   'cell',\n",
       "   'stimulate',\n",
       "   'vegf-a165',\n",
       "   'vegfr-2',\n",
       "   'phosphorylation',\n",
       "   'mitogenicity',\n",
       "   'induce',\n",
       "   'vegf-a165',\n",
       "   'enhance',\n",
       "   'cell',\n",
       "   'plate',\n",
       "   'gene_alphavbeta3_ligand_gene',\n",
       "   'gene_vitronectin_gene',\n",
       "   'compare',\n",
       "   'cell',\n",
       "   'plate',\n",
       "   'alpha5beta1',\n",
       "   'ligand',\n",
       "   'gene_fibronectin_gene',\n",
       "   'alpha2beta1',\n",
       "   'ligand',\n",
       "   'collagen',\n",
       "   'bv4',\n",
       "   'anti-beta3',\n",
       "   'integrin',\n",
       "   'mab',\n",
       "   'interfere',\n",
       "   'endothelial',\n",
       "   'cell',\n",
       "   'adhesion',\n",
       "   'gene_vitronectin_gene',\n",
       "   'reduce',\n",
       "   'i',\n",
       "   'tyrosine',\n",
       "   'phosphorylation',\n",
       "   'vegfr-2',\n",
       "   'ii',\n",
       "   'activation',\n",
       "   'downstream',\n",
       "   'transductor',\n",
       "   'phosphoinositide',\n",
       "   '3-oh',\n",
       "   'kinase',\n",
       "   'iii',\n",
       "   'biological',\n",
       "   'effect',\n",
       "   'trigger',\n",
       "   'vegf-a165',\n",
       "   'result',\n",
       "   'indicate',\n",
       "   'new',\n",
       "   'role',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'integrin',\n",
       "   'activation',\n",
       "   'vitro',\n",
       "   'angiogenic',\n",
       "   'program',\n",
       "   'endothelial',\n",
       "   'cell',\n",
       "   'besides',\n",
       "   'be',\n",
       "   'important',\n",
       "   'survival',\n",
       "   'system',\n",
       "   'nascent',\n",
       "   'vessel',\n",
       "   'regulate',\n",
       "   'cell',\n",
       "   'adhesion',\n",
       "   'matrix',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'integrin',\n",
       "   'participate',\n",
       "   'full',\n",
       "   'activation',\n",
       "   'vegfr-2',\n",
       "   'trigger',\n",
       "   'vegf-a',\n",
       "   'important',\n",
       "   'angiogenic',\n",
       "   'inducer',\n",
       "   'dis_tumor_dis',\n",
       "   'inflammation',\n",
       "   'tissue',\n",
       "   'regeneration')),\n",
       " (2,\n",
       "  ('10023672',\n",
       "   'differential',\n",
       "   'expression',\n",
       "   'translocation',\n",
       "   'protein',\n",
       "   'tyrosine',\n",
       "   'phosphatase',\n",
       "   '1b-related',\n",
       "   'protein',\n",
       "   'me-180',\n",
       "   'dis_tumor_dis',\n",
       "   'cell',\n",
       "   'express',\n",
       "   'apoptotic',\n",
       "   'sensitivity',\n",
       "   'resistance',\n",
       "   'dis_tumor_necrosis_dis',\n",
       "   'factor',\n",
       "   'potential',\n",
       "   'interaction',\n",
       "   'gene_epidermal_growth_factor_gene',\n",
       "   'receptor',\n",
       "   'dis_tumor_necrosis_dis',\n",
       "   'factor',\n",
       "   'tnf)-induced',\n",
       "   'apoptosis',\n",
       "   'inhibit',\n",
       "   'overexpression',\n",
       "   'specific',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'activation',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'cascade',\n",
       "   'suggest',\n",
       "   'potential',\n",
       "   'antagonism',\n",
       "   'between',\n",
       "   'apoptotic',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'signaling',\n",
       "   'process',\n",
       "   'report',\n",
       "   'effect',\n",
       "   'tnf',\n",
       "   'gene_egf_receptor_gene',\n",
       "   'tyrosine',\n",
       "   'phosphorylation',\n",
       "   'me-180',\n",
       "   'cell',\n",
       "   'variant',\n",
       "   'select',\n",
       "   'apoptotic',\n",
       "   'sensitivity',\n",
       "   'sen',\n",
       "   'resistance',\n",
       "   'res',\n",
       "   'tnf',\n",
       "   'previous',\n",
       "   'show',\n",
       "   'differential',\n",
       "   'express',\n",
       "   'gene_egfr+_gene',\n",
       "   'examine',\n",
       "   'prior',\n",
       "   'onset',\n",
       "   'apoptosis',\n",
       "   'tnf',\n",
       "   'cause',\n",
       "   'significant',\n",
       "   'reduction',\n",
       "   'level',\n",
       "   'gene_egfr+_gene',\n",
       "   'tyrosine',\n",
       "   'phosphorylation',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'mediate',\n",
       "   'limit',\n",
       "   'suppression',\n",
       "   'gene_egfr+_gene',\n",
       "   'tyrosine',\n",
       "   'phosphorylation',\n",
       "   'apoptotical',\n",
       "   'resistant',\n",
       "   'res',\n",
       "   'cell',\n",
       "   'vitro',\n",
       "   'incubation',\n",
       "   'cellular',\n",
       "   'membrane',\n",
       "   'tnf',\n",
       "   'derive',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'stimulate',\n",
       "   'resident',\n",
       "   'protein',\n",
       "   'tyrosine',\n",
       "   'phosphatase',\n",
       "   'ptp',\n",
       "   'activity',\n",
       "   'dephosphorylate',\n",
       "   'gene_egfr+_gene',\n",
       "   'tyrosine',\n",
       "   'phosphopeptide',\n",
       "   'mimic',\n",
       "   'gene_egfr+_gene',\n",
       "   'autophosphorylation',\n",
       "   'site',\n",
       "   'membrane',\n",
       "   'preparation',\n",
       "   'ptpib',\n",
       "   'complex',\n",
       "   'tyrosine',\n",
       "   'phosphorylate',\n",
       "   'gene_egfr+_gene',\n",
       "   'association',\n",
       "   'disrupt',\n",
       "   'tnf',\n",
       "   'through',\n",
       "   'apparent',\n",
       "   'stimulation',\n",
       "   'ptp',\n",
       "   'activity',\n",
       "   'turnover',\n",
       "   'phosphotyrosine',\n",
       "   'intrinsic',\n",
       "   'enzymatic',\n",
       "   'activity',\n",
       "   'ptp1b',\n",
       "   '2-3-fold',\n",
       "   'high',\n",
       "   'sen',\n",
       "   'versus',\n",
       "   'res',\n",
       "   'cell',\n",
       "   'lysate',\n",
       "   'family',\n",
       "   'ptp1b-related',\n",
       "   'protein',\n",
       "   'alter',\n",
       "   'c-termini',\n",
       "   'find',\n",
       "   'high',\n",
       "   'express',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'absent',\n",
       "   'express',\n",
       "   'reduce',\n",
       "   'level',\n",
       "   'res',\n",
       "   'cell',\n",
       "   'cytoplasmic',\n",
       "   'extract',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'contain',\n",
       "   'ptp1b-like',\n",
       "   'protein',\n",
       "   'tnf',\n",
       "   'incubation',\n",
       "   'result',\n",
       "   'time',\n",
       "   'dependent',\n",
       "   'accumulation',\n",
       "   'ptp1b-like',\n",
       "   'protein',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'effect',\n",
       "   'protein',\n",
       "   'res',\n",
       "   'cell',\n",
       "   'together',\n",
       "   'result',\n",
       "   'suggest',\n",
       "   'specific',\n",
       "   'change',\n",
       "   'expression',\n",
       "   'subcellular',\n",
       "   'distribution',\n",
       "   'phosphotyrosine',\n",
       "   'modulatory',\n",
       "   'protein',\n",
       "   'play',\n",
       "   'role',\n",
       "   'convey',\n",
       "   'intrinsic',\n",
       "   'apoptotic',\n",
       "   'sensitivity',\n",
       "   'tnf',\n",
       "   'dis_tumor_dis',\n",
       "   'cell',\n",
       "   'type'))]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#split_list = my_para.map(lambda x: list(x[0].split()))\n",
    "split_list = my_para.map(lambda x: tuple(x[0].split()))\n",
    "#split_list2 = split_list.filter(lambda x: x[0] is not None)\n",
    "\n",
    "\n",
    "doc_id = split_list.zipWithIndex().keyBy(lambda x: x[1])\n",
    "\n",
    "# unpacking 2 tuple into 1 singular tuple to conpute\n",
    "doc_id = doc_id.map(lambda x: tuple((x[0], *x[1])))\n",
    "\n",
    "# re-packing teh tuple\n",
    "doc_id = doc_id.map(lambda x: tuple((x[0], x[1])))\n",
    "doc_id.take(3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# separate into key-value pairs\n",
    "# ('doc_id', 'term') --> (( 'doc_id', 'token'), 1 )\n",
    "kv_mapper = doc_id.flatMap(lambda x: tuple( [ ((x[0], i),1) for i in x[1]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[((0, '10022814'), 1),\n",
       " ((0, 'growth'), 1),\n",
       " ((0, 'inhibition'), 1),\n",
       " ((0, 'dis_breast_cancer_dis'), 1),\n",
       " ((0, 'cell'), 1)]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kv_mapper.take(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[((0, '10022814'), 1),\n",
       " ((0, 'growth'), 6),\n",
       " ((0, 'inhibition'), 3),\n",
       " ((0, 'dis_breast_cancer_dis'), 7),\n",
       " ((0, 'cell'), 6),\n",
       " ((0, 'grb2'), 7),\n",
       " ((0, 'downregulation'), 2),\n",
       " ((0, 'correlate'), 1),\n",
       " ((0, 'inactivation'), 2),\n",
       " ((0, 'mitogen-activated'), 1),\n",
       " ((0, 'protein'), 3),\n",
       " ((0, 'kinase'), 4),\n",
       " ((0, 'gene_egfr+_gene'), 5),\n",
       " ((0, 'erbb2'), 5),\n",
       " ((0, 'increased'), 1),\n",
       " ((0, 'associate'), 1),\n",
       " ((0, 'increase'), 1),\n",
       " ((0, 'expression'), 1),\n",
       " ((0, 'gene_epidermal_growth_factor_gene'), 1),\n",
       " ((0, 'receptor'), 3),\n",
       " ((0, 'tyrosine'), 1),\n",
       " ((0, 'rtks'), 4),\n",
       " ((0, 'upon'), 2),\n",
       " ((0, 'activation'), 2),\n",
       " ((0, 'transmit'), 1),\n",
       " ((0, 'oncogenic'), 1),\n",
       " ((0, 'signal'), 1),\n",
       " ((0, 'binding'), 1),\n",
       " ((0, 'factor'), 1),\n",
       " ((0, 'bind'), 2),\n",
       " ((0, 'protein-2'), 1),\n",
       " ((0, 'turn'), 1),\n",
       " ((0, 'sos'), 1),\n",
       " ((0, 'activate'), 1),\n",
       " ((0, 'ras-raf-mek-mitogen-activated'), 1),\n",
       " ((0, 'map'), 2),\n",
       " ((0, 'pathway'), 2),\n",
       " ((0, 'important'), 2),\n",
       " ((0, 'transformation'), 1),\n",
       " ((0, 'fibroblast'), 1),\n",
       " ((0, 'whether'), 1),\n",
       " ((0, 'proliferation'), 1),\n",
       " ((0, 'express'), 1),\n",
       " ((0, 'unclear'), 1),\n",
       " ((0, 'use'), 2),\n",
       " ((0, 'liposome'), 1),\n",
       " ((0, 'deliver'), 1),\n",
       " ((0, 'nuclease-resistant'), 1),\n",
       " ((0, 'antisense'), 1),\n",
       " ((0, 'oligodeoxynucleotide'), 1),\n",
       " ((0, 'oligos'), 1),\n",
       " ((0, 'specific'), 1),\n",
       " ((0, 'mrna'), 1),\n",
       " ((0, 'inhibit'), 1),\n",
       " ((0, 'degree'), 1),\n",
       " ((0, 'dependent'), 1),\n",
       " ((0, 'and-or'), 1),\n",
       " ((0, 'endogenou'), 1),\n",
       " ((0, 'level'), 1),\n",
       " ((0, 'lead'), 1),\n",
       " ((0, 'suggest'), 1),\n",
       " ((0, 'different'), 1),\n",
       " ((0, 'regulate'), 1),\n",
       " ((1, '10022831'), 1),\n",
       " ((1, 'role'), 2),\n",
       " ((1, 'gene_alphavbeta3_gene'), 5),\n",
       " ((1, 'integrin'), 8),\n",
       " ((1, 'activation'), 4),\n",
       " ((1, 'vascular'), 2),\n",
       " ((1, 'endothelial'), 6),\n",
       " ((1, 'growth'), 3),\n",
       " ((1, 'factor'), 3),\n",
       " ((1, 'receptor-2'), 2),\n",
       " ((1, 'interaction'), 2),\n",
       " ((1, 'between'), 2),\n",
       " ((1, 'extracellular'), 1),\n",
       " ((1, 'matrix'), 2),\n",
       " ((1, 'crucial'), 1),\n",
       " ((1, 'cell'), 9),\n",
       " ((1, 'sprout'), 1),\n",
       " ((1, 'capillary'), 1),\n",
       " ((1, 'angiogenesis'), 1),\n",
       " ((1, 'furthermore'), 1),\n",
       " ((1, 'integrin-mediated'), 1),\n",
       " ((1, 'outside-in'), 1),\n",
       " ((1, 'signal'), 1),\n",
       " ((1, 'co-operate'), 1),\n",
       " ((1, 'receptor'), 2),\n",
       " ((1, 'promote'), 1),\n",
       " ((1, 'proliferation'), 1),\n",
       " ((1, 'motility'), 1),\n",
       " ((1, 'determine'), 1),\n",
       " ((1, 'potential'), 1),\n",
       " ((1, 'regulation'), 1),\n",
       " ((1, 'angiogenic'), 3),\n",
       " ((1, 'inducer'), 2),\n",
       " ((1, 'system'), 2),\n",
       " ((1, 'investigate'), 1),\n",
       " ((1, 'tyrosine'), 2),\n",
       " ((1, 'kinase'), 2),\n",
       " ((1, 'vegfr-2'), 5),\n",
       " ((1, 'human'), 1),\n",
       " ((1, 'report'), 1),\n",
       " ((1, 'tyrosine-phosphorylated'), 1),\n",
       " ((1, 'co-immunoprecipitated'), 1),\n",
       " ((1, 'beta3'), 1),\n",
       " ((1, 'subunit'), 1),\n",
       " ((1, 'beta1'), 1),\n",
       " ((1, 'beta5'), 1),\n",
       " ((1, 'stimulate'), 1),\n",
       " ((1, 'vegf-a165'), 3),\n",
       " ((1, 'phosphorylation'), 2),\n",
       " ((1, 'mitogenicity'), 1),\n",
       " ((1, 'induce'), 1),\n",
       " ((1, 'enhance'), 1),\n",
       " ((1, 'plate'), 2),\n",
       " ((1, 'gene_alphavbeta3_ligand_gene'), 1),\n",
       " ((1, 'gene_vitronectin_gene'), 2),\n",
       " ((1, 'compare'), 1),\n",
       " ((1, 'alpha5beta1'), 1),\n",
       " ((1, 'ligand'), 2),\n",
       " ((1, 'gene_fibronectin_gene'), 1),\n",
       " ((1, 'alpha2beta1'), 1),\n",
       " ((1, 'collagen'), 1),\n",
       " ((1, 'bv4'), 1),\n",
       " ((1, 'anti-beta3'), 1),\n",
       " ((1, 'mab'), 1),\n",
       " ((1, 'interfere'), 1),\n",
       " ((1, 'adhesion'), 2),\n",
       " ((1, 'reduce'), 1),\n",
       " ((1, 'i'), 1),\n",
       " ((1, 'ii'), 1),\n",
       " ((1, 'downstream'), 1),\n",
       " ((1, 'transductor'), 1),\n",
       " ((1, 'phosphoinositide'), 1),\n",
       " ((1, '3-oh'), 1),\n",
       " ((1, 'iii'), 1),\n",
       " ((1, 'biological'), 1),\n",
       " ((1, 'effect'), 1),\n",
       " ((1, 'trigger'), 2),\n",
       " ((1, 'result'), 1),\n",
       " ((1, 'indicate'), 1),\n",
       " ((1, 'new'), 1),\n",
       " ((1, 'vitro'), 1),\n",
       " ((1, 'program'), 1),\n",
       " ((1, 'besides'), 1),\n",
       " ((1, 'be'), 1),\n",
       " ((1, 'important'), 2),\n",
       " ((1, 'survival'), 1),\n",
       " ((1, 'nascent'), 1),\n",
       " ((1, 'vessel'), 1),\n",
       " ((1, 'regulate'), 1),\n",
       " ((1, 'participate'), 1),\n",
       " ((1, 'full'), 1),\n",
       " ((1, 'vegf-a'), 1),\n",
       " ((1, 'dis_tumor_dis'), 1),\n",
       " ((1, 'inflammation'), 1),\n",
       " ((1, 'tissue'), 1),\n",
       " ((1, 'regeneration'), 1),\n",
       " ((2, '10023672'), 1),\n",
       " ((2, 'differential'), 2),\n",
       " ((2, 'expression'), 2),\n",
       " ((2, 'translocation'), 1),\n",
       " ((2, 'protein'), 8),\n",
       " ((2, 'tyrosine'), 10),\n",
       " ((2, 'phosphatase'), 2),\n",
       " ((2, '1b-related'), 1),\n",
       " ((2, 'me-180'), 2),\n",
       " ((2, 'dis_tumor_dis'), 2),\n",
       " ((2, 'cell'), 12),\n",
       " ((2, 'express'), 4),\n",
       " ((2, 'apoptotic'), 4),\n",
       " ((2, 'sensitivity'), 3),\n",
       " ((2, 'resistance'), 2),\n",
       " ((2, 'dis_tumor_necrosis_dis'), 2),\n",
       " ((2, 'factor'), 2),\n",
       " ((2, 'potential'), 2),\n",
       " ((2, 'interaction'), 1),\n",
       " ((2, 'gene_epidermal_growth_factor_gene'), 1),\n",
       " ((2, 'receptor'), 1),\n",
       " ((2, 'tnf)-induced'), 1),\n",
       " ((2, 'apoptosis'), 2),\n",
       " ((2, 'inhibit'), 1),\n",
       " ((2, 'overexpression'), 1),\n",
       " ((2, 'specific'), 2),\n",
       " ((2, 'kinase'), 3),\n",
       " ((2, 'activation'), 1),\n",
       " ((2, 'cascade'), 1),\n",
       " ((2, 'suggest'), 2),\n",
       " ((2, 'antagonism'), 1),\n",
       " ((2, 'between'), 1),\n",
       " ((2, 'signaling'), 1),\n",
       " ((2, 'process'), 1),\n",
       " ((2, 'report'), 1),\n",
       " ((2, 'effect'), 2),\n",
       " ((2, 'tnf'), 7),\n",
       " ((2, 'gene_egf_receptor_gene'), 1),\n",
       " ((2, 'phosphorylation'), 3),\n",
       " ((2, 'variant'), 1),\n",
       " ((2, 'select'), 1),\n",
       " ((2, 'sen'), 7),\n",
       " ((2, 'res'), 5),\n",
       " ((2, 'previous'), 1),\n",
       " ((2, 'show'), 1),\n",
       " ((2, 'gene_egfr+_gene'), 6),\n",
       " ((2, 'examine'), 1),\n",
       " ((2, 'prior'), 1),\n",
       " ((2, 'onset'), 1),\n",
       " ((2, 'cause'), 1),\n",
       " ((2, 'significant'), 1),\n",
       " ((2, 'reduction'), 1),\n",
       " ((2, 'level'), 2),\n",
       " ((2, 'mediate'), 1),\n",
       " ((2, 'limit'), 1),\n",
       " ((2, 'suppression'), 1),\n",
       " ((2, 'apoptotical'), 1),\n",
       " ((2, 'resistant'), 1),\n",
       " ((2, 'vitro'), 1),\n",
       " ((2, 'incubation'), 2),\n",
       " ((2, 'cellular'), 1),\n",
       " ((2, 'membrane'), 2),\n",
       " ((2, 'derive'), 1),\n",
       " ((2, 'stimulate'), 1),\n",
       " ((2, 'resident'), 1),\n",
       " ((2, 'ptp'), 2),\n",
       " ((2, 'activity'), 3),\n",
       " ((2, 'dephosphorylate'), 1),\n",
       " ((2, 'phosphopeptide'), 1),\n",
       " ((2, 'mimic'), 1),\n",
       " ((2, 'autophosphorylation'), 1),\n",
       " ((2, 'site'), 1),\n",
       " ((2, 'preparation'), 1),\n",
       " ((2, 'ptpib'), 1),\n",
       " ((2, 'complex'), 1),\n",
       " ((2, 'phosphorylate'), 1),\n",
       " ((2, 'association'), 1),\n",
       " ((2, 'disrupt'), 1),\n",
       " ((2, 'through'), 1),\n",
       " ((2, 'apparent'), 1),\n",
       " ((2, 'stimulation'), 1),\n",
       " ((2, 'turnover'), 1),\n",
       " ((2, 'phosphotyrosine'), 2),\n",
       " ((2, 'intrinsic'), 2),\n",
       " ((2, 'enzymatic'), 1),\n",
       " ((2, 'ptp1b'), 1),\n",
       " ((2, '2-3-fold'), 1),\n",
       " ((2, 'high'), 2),\n",
       " ((2, 'versus'), 1),\n",
       " ((2, 'lysate'), 1),\n",
       " ((2, 'family'), 1),\n",
       " ((2, 'ptp1b-related'), 1),\n",
       " ((2, 'alter'), 1),\n",
       " ((2, 'c-termini'), 1),\n",
       " ((2, 'find'), 1),\n",
       " ((2, 'absent'), 1),\n",
       " ((2, 'reduce'), 1),\n",
       " ((2, 'cytoplasmic'), 1),\n",
       " ((2, 'extract'), 1),\n",
       " ((2, 'contain'), 1),\n",
       " ((2, 'ptp1b-like'), 2),\n",
       " ((2, 'result'), 2),\n",
       " ((2, 'time'), 1),\n",
       " ((2, 'dependent'), 1),\n",
       " ((2, 'accumulation'), 1),\n",
       " ((2, 'together'), 1),\n",
       " ((2, 'change'), 1),\n",
       " ((2, 'subcellular'), 1),\n",
       " ((2, 'distribution'), 1),\n",
       " ((2, 'modulatory'), 1),\n",
       " ((2, 'play'), 1),\n",
       " ((2, 'role'), 1),\n",
       " ((2, 'convey'), 1),\n",
       " ((2, 'type'), 1),\n",
       " ((3, '10029092'), 1),\n",
       " ((3, 'elevation'), 1),\n",
       " ((3, 'gene_epidermal_growth_factor_gene'), 3),\n",
       " ((3, 'receptor'), 2),\n",
       " ((3, 'dependent'), 1),\n",
       " ((3, 'signaling'), 1),\n",
       " ((3, 'human'), 3),\n",
       " ((3, 'papillomavirus-infected'), 1),\n",
       " ((3, 'laryngeal'), 9),\n",
       " ((3, 'dis_papilloma_dis'), 10),\n",
       " ((3, 'dis_benign_tumor_dis'), 1),\n",
       " ((3, 'cause'), 1),\n",
       " ((3, 'papillomavirus'), 2),\n",
       " ((3, 'type'), 1),\n",
       " ((3, '6'), 1),\n",
       " ((3, '11'), 1),\n",
       " ((3, 'study'), 1),\n",
       " ((3, 'address'), 1),\n",
       " ((3, 'alteration'), 1),\n",
       " ((3, 'level'), 2),\n",
       " ((3, 'signal'), 1),\n",
       " ((3, 'transduction'), 1),\n",
       " ((3, 'gene_egfr+_gene'), 7),\n",
       " ((3, 'culture'), 2),\n",
       " ((3, 'cell'), 9),\n",
       " ((3, 'compare'), 2),\n",
       " ((3, 'normal'), 4),\n",
       " ((3, 'tissue'), 1),\n",
       " ((3, 'mitogen-activated'), 1),\n",
       " ((3, 'protein'), 1),\n",
       " ((3, 'kinase'), 2),\n",
       " ((3, 'mapk'), 4),\n",
       " ((3, 'activate'), 1),\n",
       " ((3, 'great'), 1),\n",
       " ((3, 'extent'), 1),\n",
       " ((3, 'phosphotyrosine'), 1),\n",
       " ((3, 'many'), 1),\n",
       " ((3, 'abundant'), 2),\n",
       " ((3, 'overexpress'), 1),\n",
       " ((3, 'epithelium'), 2),\n",
       " ((3, 'western'), 2),\n",
       " ((3, 'blot'), 2),\n",
       " ((3, 'analysis'), 2),\n",
       " ((3, '3'), 1),\n",
       " ((3, 'time'), 1),\n",
       " ((3, 'much'), 2),\n",
       " ((3, 'scatchard'), 1),\n",
       " ((3, 'without'), 1),\n",
       " ((3, 'gene'), 1),\n",
       " ((3, 'amplification'), 1),\n",
       " ((3, 'increase'), 2),\n",
       " ((3, 'steady-state'), 1),\n",
       " ((3, 'mrna'), 1),\n",
       " ((3, 'following'), 1),\n",
       " ((3, 'stimulation'), 3),\n",
       " ((3, 'gene_egf_gene'), 1),\n",
       " ((3, 'significant'), 1),\n",
       " ((3, 'portion'), 1),\n",
       " ((3, 'recycled'), 1),\n",
       " ((3, 'surface'), 1),\n",
       " ((3, 'whereas'), 1),\n",
       " ((3, 'tyrosine'), 1),\n",
       " ((3, 'activity'), 1),\n",
       " ((3, 'activation'), 4),\n",
       " ((3, 'responsive'), 1),\n",
       " ((3, 'uninfect'), 1),\n",
       " ((3, 'primary'), 1),\n",
       " ((3, 'pd153035'), 1),\n",
       " ((3, 'specific'), 1),\n",
       " ((3, 'inhibitor'), 1),\n",
       " ((3, 'egfr-specific'), 1),\n",
       " ((3, 'antibody'), 1),\n",
       " ((3, 'block'), 1),\n",
       " ((3, 'ligand'), 2),\n",
       " ((3, 'binding'), 1),\n",
       " ((3, 'complete'), 1),\n",
       " ((3, 'abrogate'), 1),\n",
       " ((3, 'basal'), 1),\n",
       " ((3, 'endogenou'), 1),\n",
       " ((3, 'result'), 1),\n",
       " ((3, 'demonstrate'), 1),\n",
       " ((3, 'infection'), 1),\n",
       " ((3, 'low-risk'), 1),\n",
       " ((3, 'elevate'), 1),\n",
       " ((3, 'posttranslational'), 1),\n",
       " ((3, 'mechanism'), 1),\n",
       " ((3, 'responsiveness'), 1),\n",
       " ((3, 'ligand-mediated'), 2),\n",
       " ((3, 'show'), 1),\n",
       " ((3, 'depend'), 1),\n",
       " ((3, 'upon'), 1),\n",
       " ((4, '10029453'), 1),\n",
       " ((4, 'reverse'), 1),\n",
       " ((4, 'transcriptase'), 1),\n",
       " ((4, 'polymerase'), 2),\n",
       " ((4, 'chain'), 2),\n",
       " ((4, 'reaction'), 2),\n",
       " ((4, 'reliable'), 1),\n",
       " ((4, 'method'), 1),\n",
       " ((4, 'detect'), 1),\n",
       " ((4, 'gene_epidermal_growth_factor_gene'), 2),\n",
       " ((4, 'receptor'), 8),\n",
       " ((4, 'exon'), 3),\n",
       " ((4, '2-7'), 1),\n",
       " ((4, 'gene'), 8),\n",
       " ((4, 'deletion'), 4),\n",
       " ((4, 'human'), 1),\n",
       " ((4, 'dis_glioblastomas_dis'), 1),\n",
       " ((4, 'gene_egfr+_gene'), 7),\n",
       " ((4, 'amplification'), 5),\n",
       " ((4, 'report'), 1),\n",
       " ((4, 'occur'), 2),\n",
       " ((4, 'diverse'), 1),\n",
       " ((4, 'dis_carcinoma_dis'), 2),\n",
       " ((4, 'type'), 2),\n",
       " ((4, 'such'), 1),\n",
       " ((4, 'lung'), 1),\n",
       " ((4, 'ovarian'), 1),\n",
       " ((4, 'breast'), 1),\n",
       " ((4, 'dis_glioblastoma_dis'), 6),\n",
       " ((4, '801-bp'), 2),\n",
       " ((4, 'in-frame'), 1),\n",
       " ((4, 'close'), 1),\n",
       " ((4, 'aminoterminu'), 2),\n",
       " ((4, 'protein'), 6),\n",
       " ((4, 'find'), 2),\n",
       " ((4, 'many'), 1),\n",
       " ((4, 'little'), 1),\n",
       " ((4, 'frequent'), 2),\n",
       " ((4, 'within'), 1),\n",
       " ((4, 'three'), 1),\n",
       " ((4, 'dis_tumor_dis'), 5),\n",
       " ((4, 'entity'), 1),\n",
       " ((4, 'study'), 1),\n",
       " ((4, 'alteration'), 2),\n",
       " ((4, 'use'), 2),\n",
       " ((4, 'expression'), 3),\n",
       " ((4, '65'), 1),\n",
       " ((4, 'astrocytic'), 1),\n",
       " ((4, '51'), 2),\n",
       " ((4, 'world'), 1),\n",
       " ((4, 'health'), 1),\n",
       " ((4, 'organization'), 1),\n",
       " ((4, 'who'), 1),\n",
       " ((4, 'iv'), 1),\n",
       " ((4, 'five'), 2),\n",
       " ((4, 'dis_anaplastic_dis'), 2),\n",
       " ((4, 'dis_astrocytoma_dis'), 3),\n",
       " ((4, 'iii'), 2),\n",
       " ((4, 'nine'), 1),\n",
       " ((4, 'ii'), 1),\n",
       " ((4, 'determine'), 1),\n",
       " ((4, 'southern'), 2),\n",
       " ((4, 'blot'), 2),\n",
       " ((4, 'full-length'), 3),\n",
       " ((4, 'cdna'), 1),\n",
       " ((4, 'probe'), 1),\n",
       " ((4, 'observe'), 2),\n",
       " ((4, '22'), 1),\n",
       " ((4, '43%'), 1),\n",
       " ((4, 'none'), 1),\n",
       " ((4, 'dis_grade_ii_astrocytoma_dis'), 1),\n",
       " ((4, 'two'), 2),\n",
       " ((4, 'show'), 3),\n",
       " ((4, 'considerable'), 1),\n",
       " ((4, 'degree'), 1),\n",
       " ((4, 'accord'), 1),\n",
       " ((4, 'neuroradiological'), 1),\n",
       " ((4, 'data'), 1),\n",
       " ((4, 'consider'), 1),\n",
       " ((4, 'genetic'), 1),\n",
       " ((4, 'near'), 1),\n",
       " ((4, 'comprise'), 1),\n",
       " ((4, 'complete'), 1),\n",
       " ((4, 'loss'), 1),\n",
       " ((4, '2'), 1),\n",
       " ((4, '7'), 1),\n",
       " ((4, 'del2-7'), 1),\n",
       " ((4, 'rt-pcr'), 1),\n",
       " ((4, 'superior'), 1),\n",
       " ((4, 'analysis'), 1),\n",
       " ((4, 'detection'), 1),\n",
       " ((4, 'assign'), 1),\n",
       " ((4, '9'), 1),\n",
       " ((4, '38'), 1),\n",
       " ((4, '24%'), 1),\n",
       " ((4, 'examine'), 1),\n",
       " ((4, 'gene_egf_receptor_gene'), 1),\n",
       " ((4, 'enhance'), 1),\n",
       " ((4, '5'), 1),\n",
       " ((4, '18'), 1),\n",
       " ((4, 'express'), 1),\n",
       " ((4, 'absence'), 1),\n",
       " ((4, 'elevated'), 1),\n",
       " ((4, 'level'), 1),\n",
       " ((4, 'addition'), 1),\n",
       " ((4, 'signal'), 1),\n",
       " ((4, '140-kda'), 1),\n",
       " ((4, 'range'), 1),\n",
       " ((4, '17'), 1),\n",
       " ((4, '35'), 1),\n",
       " ((4, '49%'), 1),\n",
       " ((4, 'fragment'), 1),\n",
       " ((4, 'correspond'), 1),\n",
       " ((4, 'truncate'), 1),\n",
       " ((4, 'de12-7'), 1),\n",
       " ((4, 'due'), 1),\n",
       " ((4, 'proteolysis'), 1),\n",
       " ((5, '10037173'), 1),\n",
       " ((5, 'anti-epidermal'), 1),\n",
       " ((5, 'growth'), 8),\n",
       " ((5, 'factor'), 4),\n",
       " ((5, 'receptor'), 2),\n",
       " ((5, 'antibody'), 2),\n",
       " ((5, 'c225'), 6),\n",
       " ((5, 'inhibit'), 3),\n",
       " ((5, 'angiogenesis'), 3),\n",
       " ((5, 'human'), 5),\n",
       " ((5, 'dis_transitional_cell_carcinoma_dis'), 2),\n",
       " ((5, 'grow'), 2),\n",
       " ((5, 'orthotopical'), 4),\n",
       " ((5, 'nude'), 4),\n",
       " ((5, 'mouse'), 4),\n",
       " ((5, 'epidermal'), 1),\n",
       " ((5, 'egfr'), 2),\n",
       " ((5, 'regulate'), 1),\n",
       " ((5, 'progression'), 1),\n",
       " ((5, 'dis_tcc_dis'), 4),\n",
       " ((5, 'bladder'), 1),\n",
       " ((5, 'show'), 1),\n",
       " ((5, 'therapy'), 4),\n",
       " ((5, 'target'), 1),\n",
       " ((5, 'establish'), 3),\n",
       " ((5, 'purpose'), 1),\n",
       " ((5, 'study'), 2),\n",
       " ((5, 'evaluate'), 3),\n",
       " ((5, 'whether'), 1),\n",
       " ((5, 'egfr-directed'), 1),\n",
       " ((5, 'affect'), 1),\n",
       " ((5, 'associate'), 1),\n",
       " ((5, 'dis_metastasis_dis'), 2),\n",
       " ((5, 'determine'), 1),\n",
       " ((5, 'cytostatic'), 1),\n",
       " ((5, 'effect'), 3),\n",
       " ((5, 'production'), 2),\n",
       " ((5, 'gene_angiogenic_factor_gene'), 3),\n",
       " ((5, 'vitro'), 2),\n",
       " ((5, 'treatment'), 2),\n",
       " ((5, 'cell'), 3),\n",
       " ((5, 'line'), 1),\n",
       " ((5, '253j'), 3),\n",
       " ((5, 'b-v'), 3),\n",
       " ((5, 'mab'), 5),\n",
       " ((5, 'chimerize'), 1),\n",
       " ((5, 'monoclonal'), 1),\n",
       " ((5, 'anti-egfr'), 2),\n",
       " ((5, 'implant'), 1),\n",
       " ((5, 'athymic'), 1),\n",
       " ((5, 'dis_tumor_dis'), 2),\n",
       " ((5, '4'), 1),\n",
       " ((5, 'weeks'), 1),\n",
       " ((5, 'treat'), 2),\n",
       " ((5, 'i'), 1),\n",
       " ((5, 'p'), 2),\n",
       " ((5, 'expression'), 2),\n",
       " ((5, 'gene_vascular_gene'), 1),\n",
       " ((5, 'endothelial'), 1),\n",
       " ((5, 'vegf'), 3),\n",
       " ((5, 'interleukin-8'), 1),\n",
       " ((5, 'il-8'), 3),\n",
       " ((5, 'basic'), 1),\n",
       " ((5, 'fibroblast'), 1),\n",
       " ((5, 'bfgf'), 3),\n",
       " ((5, 'immunohistochemistry'), 1),\n",
       " ((5, 'situ'), 1),\n",
       " ((5, 'mrna'), 2),\n",
       " ((5, 'hybridization'), 1),\n",
       " ((5, 'analysis'), 1),\n",
       " ((5, 'correlate'), 1),\n",
       " ((5, 'microvessel'), 1),\n",
       " ((5, 'density'), 1),\n",
       " ((5, 'immunohistochemical'), 1),\n",
       " ((5, 'staining'), 1),\n",
       " ((5, 'anti-cd31'), 1),\n",
       " ((5, 'protein'), 1),\n",
       " ((5, 'dose-dependent'), 1),\n",
       " ((5, 'manner'), 1),\n",
       " ((5, 'tccs'), 1),\n",
       " ((5, 'result'), 1),\n",
       " ((5, 'inhibition'), 2),\n",
       " ((5, 'compare'), 1),\n",
       " ((5, 'control'), 2),\n",
       " ((5, '<0'), 1),\n",
       " ((5, '0005'), 1),\n",
       " ((5, 'significant'), 2),\n",
       " ((5, 'low'), 1),\n",
       " ((5, 'down-regulation'), 1),\n",
       " ((5, 'precede'), 1),\n",
       " ((5, 'involution'), 1),\n",
       " ((5, 'blood'), 1),\n",
       " ((5, 'vessel'), 1),\n",
       " ((5, 'indicate'), 1),\n",
       " ((5, 'antitumor'), 1),\n",
       " ((5, 'mediated'), 1),\n",
       " ((5, 'part'), 1),\n",
       " ((6, '10037190'), 1),\n",
       " ((6, 'inducible'), 2),\n",
       " ((6, 'expression'), 5),\n",
       " ((6, 'dominant-negative'), 2),\n",
       " ((6, 'gene_epidermal_growth_factor_gene'), 2),\n",
       " ((6, 'receptor-cd533'), 1),\n",
       " ((6, 'result'), 2),\n",
       " ((6, 'radiosensitization'), 2),\n",
       " ((6, 'human'), 1),\n",
       " ((6, 'dis_mammary_carcinoma_dis'), 2),\n",
       " ((6, 'cell'), 6),\n",
       " ((6, 'ionizing'), 1),\n",
       " ((6, 'radiation'), 4),\n",
       " ((6, 'activate'), 1),\n",
       " ((6, 'receptor'), 2),\n",
       " ((6, 'gene_egfr+_gene'), 6),\n",
       " ((6, 'downstream'), 1),\n",
       " ((6, 'signaling'), 1),\n",
       " ((6, 'involve'), 1),\n",
       " ((6, 'cytoprotective'), 2),\n",
       " ((6, 'mitogen-activated'), 1),\n",
       " ((6, 'protein'), 1),\n",
       " ((6, 'kinase'), 2),\n",
       " ((6, 'mapk'), 2),\n",
       " ((6, 'pathway'), 1),\n",
       " ((6, 'effort'), 1),\n",
       " ((6, 'investigate'), 1),\n",
       " ((6, 'role'), 1),\n",
       " ((6, 'cellular'), 1),\n",
       " ((6, 'response'), 2),\n",
       " ((6, 'generate'), 1),\n",
       " ((6, 'clone'), 1),\n",
       " ((6, 'mcf-tr5-egfr-cd533'), 1),\n",
       " ((6, 'mda-tr15-egfr-cd533'), 1),\n",
       " ((6, 'express'), 1),\n",
       " ((6, 'egfr-cd533'), 4),\n",
       " ((6, 'truncate'), 1),\n",
       " ((6, 'mutant'), 2),\n",
       " ((6, 'lack'), 1),\n",
       " ((6, 'mitogenic'), 1),\n",
       " ((6, 'transformation'), 1),\n",
       " ((6, 'activity'), 1),\n",
       " ((6, 'inhibit'), 2),\n",
       " ((6, 'radiation-'), 1),\n",
       " ((6, 'egf-induced'), 1),\n",
       " ((6, 'autophosphorylation'), 1),\n",
       " ((6, 'activation'), 1),\n",
       " ((6, 'therefore'), 1),\n",
       " ((6, 'function'), 2),\n",
       " ((6, 'without'), 1),\n",
       " ((6, 'block'), 1),\n",
       " ((6, 'erbb-2'), 1),\n",
       " ((6, 'another'), 1),\n",
       " ((6, 'member'), 1),\n",
       " ((6, 'erbb'), 1),\n",
       " ((6, 'tyr'), 1),\n",
       " ((6, 'family'), 1),\n",
       " ((6, 'minimal'), 1),\n",
       " ((6, 'growth'), 2),\n",
       " ((6, 'alter'), 1),\n",
       " ((6, 'radiosensitivity'), 1),\n",
       " ((6, 'single'), 1),\n",
       " ((6, 'exposure'), 2),\n",
       " ((6, 'repeat'), 1),\n",
       " ((6, '2'), 1),\n",
       " ((6, 'gy'), 1),\n",
       " ((6, 'under'), 1),\n",
       " ((6, 'condition'), 1),\n",
       " ((6, 'essential'), 1),\n",
       " ((6, 'abolish'), 1),\n",
       " ((6, 'ability'), 1),\n",
       " ((6, 'subsequent'), 1),\n",
       " ((6, 'identify'), 1),\n",
       " ((6, 'inhibition'), 1),\n",
       " ((6, 'through'), 1),\n",
       " ((6, 'genetic'), 1),\n",
       " ((6, 'manipulation'), 1),\n",
       " ((6, 'potential'), 1),\n",
       " ((6, 'therapeutic'), 1),\n",
       " ((6, 'maneuver'), 1),\n",
       " ((6, 'concept'), 1),\n",
       " ((6, 'such'), 1),\n",
       " ((6, 'intervention'), 1),\n",
       " ((6, 'counteract'), 1),\n",
       " ((6, 'radiation-induced'), 1),\n",
       " ((6, 'proliferation'), 1),\n",
       " ((7, '10048712'), 1),\n",
       " ((7, 'molecular'), 1),\n",
       " ((7, 'genetic'), 3),\n",
       " ((7, 'characterisation'), 2),\n",
       " ((7, 'intracerebral'), 1),\n",
       " ((7, 'transplant'), 1),\n",
       " ((7, 'dis_brain_tumour_dis'), 1),\n",
       " ((7, 'aim'), 1),\n",
       " ((7, 'present'), 1),\n",
       " ((7, 'study'), 1),\n",
       " ((7, 'alteration'), 2),\n",
       " ((7, 'two'), 1),\n",
       " ((7, 'different'), 1),\n",
       " ((7, 'dis_experimental_glioma_dis'), 1),\n",
       " ((7, 'induce'), 1),\n",
       " ((7, 'rat'), 2),\n",
       " ((7, 'inbr'), 1),\n",
       " ((7, 'strain'), 1),\n",
       " ((7, 'bdix'), 1),\n",
       " ((7, 'transplacental'), 1),\n",
       " ((7, 'ethylnitrosourea'), 1),\n",
       " ((7, 'subsequent'), 1),\n",
       " ((7, 'serial'), 1),\n",
       " ((7, 'transplantation'), 1),\n",
       " ((7, 'gene'), 6),\n",
       " ((7, 'investigate'), 1),\n",
       " ((7, 'show'), 1),\n",
       " ((7, 'previous'), 1),\n",
       " ((7, 'alter'), 1),\n",
       " ((7, 'during'), 1),\n",
       " ((7, 'human'), 2),\n",
       " ((7, 'glial'), 2),\n",
       " ((7, 'dis_tumour_dis'), 3),\n",
       " ((7, 'progression'), 1),\n",
       " ((7, 'include'), 1),\n",
       " ((7, 'epidermal'), 1),\n",
       " ((7, 'growth'), 1),\n",
       " ((7, 'factor'), 1),\n",
       " ((7, 'receptor'), 1),\n",
       " ((7, 'egfr'), 1),\n",
       " ((7, 'cell'), 1),\n",
       " ((7, 'cycle'), 1),\n",
       " ((7, 'regulator'), 1),\n",
       " ((7, 'gene_cyclin_dependent_kinase_4_gene'), 1),\n",
       " ((7, 'cdk4'), 1),\n",
       " ((7, 'cyclind1'), 1),\n",
       " ((7, 'cycd1'), 1),\n",
       " ((7, 'p16'), 1),\n",
       " ((7, 'mts1-ink4'), 1),\n",
       " ((7, 'retinoblastoma_gene'), 1),\n",
       " ((7, 'dis_rb_dis'), 1),\n",
       " ((7, 'using'), 1),\n",
       " ((7, 'semi-quantitative'), 1),\n",
       " ((7, 'pcr-based'), 1),\n",
       " ((7, 'screening'), 1),\n",
       " ((7, 'method'), 1),\n",
       " ((7, 'gross'), 1),\n",
       " ((7, 'detect'), 1),\n",
       " ((7, 'demonstrate'), 1),\n",
       " ((7, 'nitrosourea-induced'), 1),\n",
       " ((7, 'harbour'), 1),\n",
       " ((7, 'those'), 1),\n",
       " ((7, 'change'), 1),\n",
       " ((7, 'typical'), 1),\n",
       " ((7, 'arise'), 1),\n",
       " ((7, 'dis_malignant_glioma_dis'), 1),\n",
       " ((7, 'thus'), 1),\n",
       " ((7, 'use'), 1),\n",
       " ((7, 'model'), 1),\n",
       " ((7, 'therapy'), 1),\n",
       " ((7, 'trial'), 1),\n",
       " ((7, 'questionable'), 1),\n",
       " ((8, '10064336'), 1),\n",
       " ((8, 'peroxisome'), 2),\n",
       " ((8, 'proliferator-activated'), 2),\n",
       " ((8, 'receptor'), 4),\n",
       " ((8, 'critical'), 1),\n",
       " ((8, 'link'), 1),\n",
       " ((8, 'fatty_acid'), 9),\n",
       " ((8, 'gene'), 2),\n",
       " ((8, 'expression'), 2),\n",
       " ((8, 'carcinogenesis'), 1),\n",
       " ((8, 'know'), 2),\n",
       " ((8, 'many'), 1),\n",
       " ((8, 'year'), 1),\n",
       " ((8, 'long-chain'), 1),\n",
       " ((8, 'derive'), 1),\n",
       " ((8, 'endogenou'), 1),\n",
       " ((8, 'metabolism'), 1),\n",
       " ((8, 'and-or'), 1),\n",
       " ((8, 'nutrition'), 1),\n",
       " ((8, 'act'), 1),\n",
       " ((8, 'second'), 2),\n",
       " ((8, 'messenger'), 2),\n",
       " ((8, 'regulator'), 1),\n",
       " ((8, 'cell'), 2),\n",
       " ((8, 'signal'), 3),\n",
       " ((8, 'pathway'), 3),\n",
       " ((8, 'example'), 1),\n",
       " ((8, 'regulate'), 1),\n",
       " ((8, 'activity'), 2),\n",
       " ((8, 'protein'), 2),\n",
       " ((8, 'kinase'), 1),\n",
       " ((8, 'c'), 1),\n",
       " ((8, 'dis_pkc_dis'), 3),\n",
       " ((8, 'mechanism'), 1),\n",
       " ((8, 'distinct'), 2),\n",
       " ((8, 'activation'), 1),\n",
       " ((8, 'diacylglycerol'), 1),\n",
       " ((8, 'activator'), 1),\n",
       " ((8, 'such'), 3),\n",
       " ((8, 'phorbol'), 1),\n",
       " ((8, 'ester'), 1),\n",
       " ((8, 'essential'), 1),\n",
       " ((8, 'activate'), 3),\n",
       " ((8, 'do'), 1),\n",
       " ((8, 'modulate'), 2),\n",
       " ((8, 'growth'), 2),\n",
       " ((8, 'factor'), 1),\n",
       " ((8, 'dis_epidermal_growth_factor_dis'), 1),\n",
       " ((8, 'egfr'), 1),\n",
       " ((8, 'unsaturated'), 1),\n",
       " ((8, 'inhibit'), 1),\n",
       " ((8, 'gtpase'), 1),\n",
       " ((8, 'thereby'), 1),\n",
       " ((8, 'quench'), 1),\n",
       " ((8, 'p21-ras'), 1),\n",
       " ((8, 'study'), 1),\n",
       " ((8, 'show'), 2),\n",
       " ((8, 'influence'), 1),\n",
       " ((8, 'numerous'), 1),\n",
       " ((8, 'signaling'), 1),\n",
       " ((8, 'small'), 1),\n",
       " ((8, 'lipophilic'), 1),\n",
       " ((8, 'substance'), 1),\n",
       " ((8, 'ancient'), 1),\n",
       " ((8, 'modulator'), 1),\n",
       " ((8, 'carcinogenic'), 1),\n",
       " ((8, 'process'), 2),\n",
       " ((8, 'tissue-specific'), 2),\n",
       " ((8, 'pro-'), 1),\n",
       " ((8, 'anticancer'), 1),\n",
       " ((8, 'effect'), 4),\n",
       " ((8, 'reason'), 1),\n",
       " ((8, 'dichotomou'), 1),\n",
       " ((8, 'cellular'), 1),\n",
       " ((8, 'adequate'), 1),\n",
       " ((8, 'describe'), 2),\n",
       " ((8, 'article'), 1),\n",
       " ((8, 'inclusion'), 1),\n",
       " ((8, 'steroid'), 1),\n",
       " ((8, 'hormone'), 1),\n",
       " ((8, 'receptor-signaling'), 1),\n",
       " ((8, 'mediate'), 2),\n",
       " ((8, 'fatty'), 2),\n",
       " ((8, \"acids'\"), 1),\n",
       " ((8, 'summarize'), 1),\n",
       " ((8, 'molecule'), 1),\n",
       " ((8, 'deem'), 1),\n",
       " ((8, 'ppar'), 3),\n",
       " ((8, 'extensive'), 1),\n",
       " ((8, 'examine'), 1),\n",
       " ((8, 'regard'), 1),\n",
       " ((8, 'response'), 2),\n",
       " ((8, 'xenobiotic'), 1),\n",
       " ((8, 'acid-like'), 1),\n",
       " ((8, 'chemical'), 1),\n",
       " ((8, 'gene_peroxisome_proliferators_gene'), 1),\n",
       " ((8, 'pp'), 2),\n",
       " ((8, 'goal'), 1),\n",
       " ((8, 'review'), 1),\n",
       " ((8, 'potential'), 1),\n",
       " ((8, 'role'), 1),\n",
       " ((8, 'differentiation'), 1),\n",
       " ((8, 'apoptosis'), 1),\n",
       " ((9, '10068212'), 1),\n",
       " ((9, 'expression'), 4),\n",
       " ((9, 'gene_vascular_endothelial_growth_factor_receptor_3_gene'), 2),\n",
       " ((9, 'podoplanin'), 4),\n",
       " ((9, 'suggest'), 2),\n",
       " ((9, 'lymphatic'), 4),\n",
       " ((9, 'endothelial'), 3),\n",
       " ((9, 'cell'), 16),\n",
       " ((9, 'origin'), 1),\n",
       " ((9, 'dis_kaposi_sarcoma_tumor_dis'), 1),\n",
       " ((9, 'despite'), 1),\n",
       " ((9, 'intensive'), 1),\n",
       " ((9, 'research'), 1),\n",
       " ((9, 'over'), 1),\n",
       " ((9, 'past'), 1),\n",
       " ((9, 'decade'), 1),\n",
       " ((9, 'exact'), 1),\n",
       " ((9, 'lineage'), 2),\n",
       " ((9, 'relationship'), 1),\n",
       " ((9, 'dis_kaposi_sarcoma_dis'), 1),\n",
       " ((9, 'dis_ks_tumor_dis'), 4),\n",
       " ((9, 'settle'), 1),\n",
       " ((9, 'present'), 2),\n",
       " ((9, 'study'), 1),\n",
       " ((9, 'investigate'), 1),\n",
       " ((9, 'two'), 2),\n",
       " ((9, 'marker'), 3),\n",
       " ((9, 'ec'), 3),\n",
       " ((9, 'ie'), 1),\n",
       " ((9, 'vegfr-3'), 3),\n",
       " ((9, 'dis_aids_dis'), 1),\n",
       " ((9, 'classic'), 1),\n",
       " ((9, 'ks'), 6),\n",
       " ((9, 'both'), 2),\n",
       " ((9, 'strong'), 2),\n",
       " ((9, 'express'), 3),\n",
       " ((9, 'line'), 2),\n",
       " ((9, 'irregular'), 1),\n",
       " ((9, 'vascular'), 1),\n",
       " ((9, 'space'), 1),\n",
       " ((9, 'early'), 1),\n",
       " ((9, 'lesion'), 3),\n",
       " ((9, 'dis_tumor_dis'), 2),\n",
       " ((9, 'advance'), 1),\n",
       " ((9, 'double-staining'), 1),\n",
       " ((9, 'experiment'), 1),\n",
       " ((9, 'confocal'), 1),\n",
       " ((9, 'laser'), 1),\n",
       " ((9, 'microscopy'), 1),\n",
       " ((9, 'establish'), 1),\n",
       " ((9, 'vegfr-3-positive'), 1),\n",
       " ((9, 'podoplanin-positive'), 1),\n",
       " ((9, 'population'), 1),\n",
       " ((9, 'identical'), 1),\n",
       " ((9, 'uniform'), 1),\n",
       " ((9, 'cd31'), 1),\n",
       " ((9, 'contrast'), 2),\n",
       " ((9, 'negative'), 1),\n",
       " ((9, 'cd45'), 1),\n",
       " ((9, 'cd68'), 1),\n",
       " ((9, 'pal-e'), 1),\n",
       " ((9, 'exclude'), 1),\n",
       " ((9, 'hemopoietic'), 1),\n",
       " ((9, 'blood'), 1),\n",
       " ((9, 'vessel'), 1),\n",
       " ((9, 'nature'), 1),\n",
       " ((9, 'primary'), 1),\n",
       " ((9, 'lysate'), 1),\n",
       " ((9, 'confirm'), 1),\n",
       " ((9, 'western'), 1),\n",
       " ((9, 'blot'), 1),\n",
       " ((9, 'analysis'), 1),\n",
       " ((9, 'splice'), 1),\n",
       " ((9, 'variant'), 1),\n",
       " ((9, 'find'), 1),\n",
       " ((9, 'ks-tumor-derived'), 1),\n",
       " ((9, 'rna'), 1),\n",
       " ((9, 'rt-pcr'), 1),\n",
       " ((9, 'situ'), 2),\n",
       " ((9, 'detect'), 1),\n",
       " ((9, 'four'), 1),\n",
       " ((9, 'ks-derived'), 2),\n",
       " ((9, 'spindle'), 1),\n",
       " ((9, 'culture'), 3),\n",
       " ((9, 'one'), 1),\n",
       " ((9, 'autonomous'), 1),\n",
       " ((9, 'grow'), 2),\n",
       " ((9, 'y-1'), 1),\n",
       " ((9, 'finding'), 1),\n",
       " ((9, 'relate'), 1),\n",
       " ((9, 'even'), 1),\n",
       " ((9, 'derive'), 2),\n",
       " ((9, 'lack'), 1),\n",
       " ((9, 'antigen'), 1),\n",
       " ((9, 'indicate'), 1),\n",
       " ((9, 'represent'), 1),\n",
       " ((9, 'tissue'), 1),\n",
       " ((9, 'type'), 1),\n",
       " ((10, '10069594'), 1),\n",
       " ((10, 'inhibition'), 3),\n",
       " ((10, 'phospholipase'), 2),\n",
       " ((10, 'c-gamma1'), 1),\n",
       " ((10, 'activation'), 7),\n",
       " ((10, 'dis_block_glioma_dis'), 1),\n",
       " ((10, 'cell'), 7),\n",
       " ((10, 'motility'), 4),\n",
       " ((10, 'invasion'), 5),\n",
       " ((10, 'fetal'), 3),\n",
       " ((10, 'rat'), 4),\n",
       " ((10, 'brain'), 3),\n",
       " ((10, 'aggregate'), 2),\n",
       " ((10, 'c'), 1),\n",
       " ((10, 'plc)-gamma'), 1),\n",
       " ((10, 'cytosolic'), 1),\n",
       " ((10, 'enzyme'), 1),\n",
       " ((10, 'activate'), 1),\n",
       " ((10, 'several'), 1),\n",
       " ((10, 'growth'), 1),\n",
       " ((10, 'factor'), 1),\n",
       " ((10, 'gf'), 8),\n",
       " ((10, 'receptor'), 5),\n",
       " ((10, 'epidermal'), 2),\n",
       " ((10, 'egfr'), 4),\n",
       " ((10, 'platelet-derived'), 2),\n",
       " ((10, 'insulin-like'), 2),\n",
       " ((10, '1'), 2),\n",
       " ((10, 'associate'), 1),\n",
       " ((10, 'increase'), 1),\n",
       " ((10, 'but'), 1),\n",
       " ((10, 'proliferation'), 1),\n",
       " ((10, 'nonglioma'), 1),\n",
       " ((10, 'line'), 2),\n",
       " ((10, 'up-regulated'), 1),\n",
       " ((10, 'consistent'), 1),\n",
       " ((10, 'link'), 1),\n",
       " ((10, 'poor'), 1),\n",
       " ((10, 'patient'), 4),\n",
       " ((10, 'survival'), 1),\n",
       " ((10, 'dis_glioblastoma_dis'), 1),\n",
       " ((10, 'multiforme'), 1),\n",
       " ((10, 'dis_gbm_dis'), 3),\n",
       " ((10, 'tyrosine'), 2),\n",
       " ((10, 'kinase'), 1),\n",
       " ((10, 'inhibitor'), 2),\n",
       " ((10, 'prevent'), 1),\n",
       " ((10, 'dis_glioma_dis'), 2),\n",
       " ((10, 'infiltration'), 3),\n",
       " ((10, 'vitro'), 2),\n",
       " ((10, 'hypothesise'), 1),\n",
       " ((10, 'gene_plc_gamma_gene'), 2),\n",
       " ((10, 'inhibit'), 1),\n",
       " ((10, 'invasiveness'), 2),\n",
       " ((10, 'experimental'), 1),\n",
       " ((10, 'model'), 2),\n",
       " ((10, 'assess'), 1),\n",
       " ((10, 'dis_tumor_dis'), 2),\n",
       " ((10, 'spheroid'), 5),\n",
       " ((10, 'confocal'), 2),\n",
       " ((10, 'microscopy'), 2),\n",
       " ((10, 'treat'), 1),\n",
       " ((10, 'gene_u87_gene'), 2),\n",
       " ...]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# do the term-count\n",
    "# (('doc_id', 'token'), 1) --> (('doc_id', 'token'), [1++])\n",
    "reduce_1 = kv_mapper.reduceByKey(lambda x, y: x+y)\n",
    "reduce_1.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('10022814', (0, 1))]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mapping for IDF start\n",
    "# (('doc_id', 'token'), tf) --> ('token', ('doc_id', tf))\n",
    "shuffle = reduce_1.map(lambda x: (x[0][1], (x[0][0], x[1])))\n",
    "shuffle.take(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('10022814', (0, 1, 1)), ('growth', (0, 6, 1))]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# tokenise for idf\n",
    "# (('doc_id', 'token'), tf) --> ('token', ('doc_id', tf, 1))\n",
    "mapper_2 = reduce_1.map(lambda x: (x[0][1], (x[0][0], x[1],1)))\n",
    "mapper_2.take(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('10022814', 1), ('growth', 1), ('inhibition', 1)]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# only take term and token\n",
    "# ('token', ('doc_id', tf,1)) --> ('token', 1)\n",
    "mapper_3 = mapper_2.map(lambda x: (x[0], x[1][2]))\n",
    "mapper_3.take(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0,\n",
       "  ('10022814',\n",
       "   'growth',\n",
       "   'inhibition',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'grb2',\n",
       "   'downregulation',\n",
       "   'correlate',\n",
       "   'inactivation',\n",
       "   'mitogen-activated',\n",
       "   'protein',\n",
       "   'kinase',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'cell',\n",
       "   'increased',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'growth',\n",
       "   'associate',\n",
       "   'increase',\n",
       "   'expression',\n",
       "   'gene_epidermal_growth_factor_gene',\n",
       "   'receptor',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'receptor',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'rtks',\n",
       "   'upon',\n",
       "   'activation',\n",
       "   'rtks',\n",
       "   'transmit',\n",
       "   'oncogenic',\n",
       "   'signal',\n",
       "   'binding',\n",
       "   'growth',\n",
       "   'factor',\n",
       "   'receptor',\n",
       "   'bind',\n",
       "   'protein-2',\n",
       "   'grb2',\n",
       "   'turn',\n",
       "   'bind',\n",
       "   'sos',\n",
       "   'activate',\n",
       "   'ras-raf-mek-mitogen-activated',\n",
       "   'protein',\n",
       "   'map',\n",
       "   'kinase',\n",
       "   'pathway',\n",
       "   'grb2',\n",
       "   'important',\n",
       "   'transformation',\n",
       "   'fibroblast',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'whether',\n",
       "   'grb2',\n",
       "   'important',\n",
       "   'proliferation',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'express',\n",
       "   'rtks',\n",
       "   'unclear',\n",
       "   'use',\n",
       "   'liposome',\n",
       "   'deliver',\n",
       "   'nuclease-resistant',\n",
       "   'antisense',\n",
       "   'oligodeoxynucleotide',\n",
       "   'oligos',\n",
       "   'specific',\n",
       "   'grb2',\n",
       "   'mrna',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'grb2',\n",
       "   'protein',\n",
       "   'downregulation',\n",
       "   'inhibit',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'growth',\n",
       "   'degree',\n",
       "   'growth',\n",
       "   'inhibition',\n",
       "   'dependent',\n",
       "   'upon',\n",
       "   'activation',\n",
       "   'and-or',\n",
       "   'endogenou',\n",
       "   'level',\n",
       "   'rtks',\n",
       "   'grb2',\n",
       "   'inhibition',\n",
       "   'lead',\n",
       "   'map',\n",
       "   'kinase',\n",
       "   'inactivation',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'cell',\n",
       "   'suggest',\n",
       "   'different',\n",
       "   'pathway',\n",
       "   'use',\n",
       "   'gene_egfr+_gene',\n",
       "   'erbb2',\n",
       "   'regulate',\n",
       "   'dis_breast_cancer_dis',\n",
       "   'growth')),\n",
       " (1,\n",
       "  ('10022831',\n",
       "   'role',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'integrin',\n",
       "   'activation',\n",
       "   'vascular',\n",
       "   'endothelial',\n",
       "   'growth',\n",
       "   'factor',\n",
       "   'receptor-2',\n",
       "   'interaction',\n",
       "   'between',\n",
       "   'integrin',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'extracellular',\n",
       "   'matrix',\n",
       "   'crucial',\n",
       "   'endothelial',\n",
       "   'cell',\n",
       "   'sprout',\n",
       "   'capillary',\n",
       "   'angiogenesis',\n",
       "   'furthermore',\n",
       "   'integrin-mediated',\n",
       "   'outside-in',\n",
       "   'signal',\n",
       "   'co-operate',\n",
       "   'growth',\n",
       "   'factor',\n",
       "   'receptor',\n",
       "   'promote',\n",
       "   'cell',\n",
       "   'proliferation',\n",
       "   'motility',\n",
       "   'determine',\n",
       "   'potential',\n",
       "   'regulation',\n",
       "   'angiogenic',\n",
       "   'inducer',\n",
       "   'receptor',\n",
       "   'integrin',\n",
       "   'system',\n",
       "   'investigate',\n",
       "   'interaction',\n",
       "   'between',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'integrin',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'vascular',\n",
       "   'endothelial',\n",
       "   'growth',\n",
       "   'factor',\n",
       "   'receptor-2',\n",
       "   'vegfr-2',\n",
       "   'human',\n",
       "   'endothelial',\n",
       "   'cell',\n",
       "   'report',\n",
       "   'tyrosine-phosphorylated',\n",
       "   'vegfr-2',\n",
       "   'co-immunoprecipitated',\n",
       "   'beta3',\n",
       "   'integrin',\n",
       "   'subunit',\n",
       "   'beta1',\n",
       "   'beta5',\n",
       "   'cell',\n",
       "   'stimulate',\n",
       "   'vegf-a165',\n",
       "   'vegfr-2',\n",
       "   'phosphorylation',\n",
       "   'mitogenicity',\n",
       "   'induce',\n",
       "   'vegf-a165',\n",
       "   'enhance',\n",
       "   'cell',\n",
       "   'plate',\n",
       "   'gene_alphavbeta3_ligand_gene',\n",
       "   'gene_vitronectin_gene',\n",
       "   'compare',\n",
       "   'cell',\n",
       "   'plate',\n",
       "   'alpha5beta1',\n",
       "   'ligand',\n",
       "   'gene_fibronectin_gene',\n",
       "   'alpha2beta1',\n",
       "   'ligand',\n",
       "   'collagen',\n",
       "   'bv4',\n",
       "   'anti-beta3',\n",
       "   'integrin',\n",
       "   'mab',\n",
       "   'interfere',\n",
       "   'endothelial',\n",
       "   'cell',\n",
       "   'adhesion',\n",
       "   'gene_vitronectin_gene',\n",
       "   'reduce',\n",
       "   'i',\n",
       "   'tyrosine',\n",
       "   'phosphorylation',\n",
       "   'vegfr-2',\n",
       "   'ii',\n",
       "   'activation',\n",
       "   'downstream',\n",
       "   'transductor',\n",
       "   'phosphoinositide',\n",
       "   '3-oh',\n",
       "   'kinase',\n",
       "   'iii',\n",
       "   'biological',\n",
       "   'effect',\n",
       "   'trigger',\n",
       "   'vegf-a165',\n",
       "   'result',\n",
       "   'indicate',\n",
       "   'new',\n",
       "   'role',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'integrin',\n",
       "   'activation',\n",
       "   'vitro',\n",
       "   'angiogenic',\n",
       "   'program',\n",
       "   'endothelial',\n",
       "   'cell',\n",
       "   'besides',\n",
       "   'be',\n",
       "   'important',\n",
       "   'survival',\n",
       "   'system',\n",
       "   'nascent',\n",
       "   'vessel',\n",
       "   'regulate',\n",
       "   'cell',\n",
       "   'adhesion',\n",
       "   'matrix',\n",
       "   'gene_alphavbeta3_gene',\n",
       "   'integrin',\n",
       "   'participate',\n",
       "   'full',\n",
       "   'activation',\n",
       "   'vegfr-2',\n",
       "   'trigger',\n",
       "   'vegf-a',\n",
       "   'important',\n",
       "   'angiogenic',\n",
       "   'inducer',\n",
       "   'dis_tumor_dis',\n",
       "   'inflammation',\n",
       "   'tissue',\n",
       "   'regeneration')),\n",
       " (2,\n",
       "  ('10023672',\n",
       "   'differential',\n",
       "   'expression',\n",
       "   'translocation',\n",
       "   'protein',\n",
       "   'tyrosine',\n",
       "   'phosphatase',\n",
       "   '1b-related',\n",
       "   'protein',\n",
       "   'me-180',\n",
       "   'dis_tumor_dis',\n",
       "   'cell',\n",
       "   'express',\n",
       "   'apoptotic',\n",
       "   'sensitivity',\n",
       "   'resistance',\n",
       "   'dis_tumor_necrosis_dis',\n",
       "   'factor',\n",
       "   'potential',\n",
       "   'interaction',\n",
       "   'gene_epidermal_growth_factor_gene',\n",
       "   'receptor',\n",
       "   'dis_tumor_necrosis_dis',\n",
       "   'factor',\n",
       "   'tnf)-induced',\n",
       "   'apoptosis',\n",
       "   'inhibit',\n",
       "   'overexpression',\n",
       "   'specific',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'activation',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'cascade',\n",
       "   'suggest',\n",
       "   'potential',\n",
       "   'antagonism',\n",
       "   'between',\n",
       "   'apoptotic',\n",
       "   'tyrosine',\n",
       "   'kinase',\n",
       "   'signaling',\n",
       "   'process',\n",
       "   'report',\n",
       "   'effect',\n",
       "   'tnf',\n",
       "   'gene_egf_receptor_gene',\n",
       "   'tyrosine',\n",
       "   'phosphorylation',\n",
       "   'me-180',\n",
       "   'cell',\n",
       "   'variant',\n",
       "   'select',\n",
       "   'apoptotic',\n",
       "   'sensitivity',\n",
       "   'sen',\n",
       "   'resistance',\n",
       "   'res',\n",
       "   'tnf',\n",
       "   'previous',\n",
       "   'show',\n",
       "   'differential',\n",
       "   'express',\n",
       "   'gene_egfr+_gene',\n",
       "   'examine',\n",
       "   'prior',\n",
       "   'onset',\n",
       "   'apoptosis',\n",
       "   'tnf',\n",
       "   'cause',\n",
       "   'significant',\n",
       "   'reduction',\n",
       "   'level',\n",
       "   'gene_egfr+_gene',\n",
       "   'tyrosine',\n",
       "   'phosphorylation',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'mediate',\n",
       "   'limit',\n",
       "   'suppression',\n",
       "   'gene_egfr+_gene',\n",
       "   'tyrosine',\n",
       "   'phosphorylation',\n",
       "   'apoptotical',\n",
       "   'resistant',\n",
       "   'res',\n",
       "   'cell',\n",
       "   'vitro',\n",
       "   'incubation',\n",
       "   'cellular',\n",
       "   'membrane',\n",
       "   'tnf',\n",
       "   'derive',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'stimulate',\n",
       "   'resident',\n",
       "   'protein',\n",
       "   'tyrosine',\n",
       "   'phosphatase',\n",
       "   'ptp',\n",
       "   'activity',\n",
       "   'dephosphorylate',\n",
       "   'gene_egfr+_gene',\n",
       "   'tyrosine',\n",
       "   'phosphopeptide',\n",
       "   'mimic',\n",
       "   'gene_egfr+_gene',\n",
       "   'autophosphorylation',\n",
       "   'site',\n",
       "   'membrane',\n",
       "   'preparation',\n",
       "   'ptpib',\n",
       "   'complex',\n",
       "   'tyrosine',\n",
       "   'phosphorylate',\n",
       "   'gene_egfr+_gene',\n",
       "   'association',\n",
       "   'disrupt',\n",
       "   'tnf',\n",
       "   'through',\n",
       "   'apparent',\n",
       "   'stimulation',\n",
       "   'ptp',\n",
       "   'activity',\n",
       "   'turnover',\n",
       "   'phosphotyrosine',\n",
       "   'intrinsic',\n",
       "   'enzymatic',\n",
       "   'activity',\n",
       "   'ptp1b',\n",
       "   '2-3-fold',\n",
       "   'high',\n",
       "   'sen',\n",
       "   'versus',\n",
       "   'res',\n",
       "   'cell',\n",
       "   'lysate',\n",
       "   'family',\n",
       "   'ptp1b-related',\n",
       "   'protein',\n",
       "   'alter',\n",
       "   'c-termini',\n",
       "   'find',\n",
       "   'high',\n",
       "   'express',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'absent',\n",
       "   'express',\n",
       "   'reduce',\n",
       "   'level',\n",
       "   'res',\n",
       "   'cell',\n",
       "   'cytoplasmic',\n",
       "   'extract',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'contain',\n",
       "   'ptp1b-like',\n",
       "   'protein',\n",
       "   'tnf',\n",
       "   'incubation',\n",
       "   'result',\n",
       "   'time',\n",
       "   'dependent',\n",
       "   'accumulation',\n",
       "   'ptp1b-like',\n",
       "   'protein',\n",
       "   'sen',\n",
       "   'cell',\n",
       "   'effect',\n",
       "   'protein',\n",
       "   'res',\n",
       "   'cell',\n",
       "   'together',\n",
       "   'result',\n",
       "   'suggest',\n",
       "   'specific',\n",
       "   'change',\n",
       "   'expression',\n",
       "   'subcellular',\n",
       "   'distribution',\n",
       "   'phosphotyrosine',\n",
       "   'modulatory',\n",
       "   'protein',\n",
       "   'play',\n",
       "   'role',\n",
       "   'convey',\n",
       "   'intrinsic',\n",
       "   'apoptotic',\n",
       "   'sensitivity',\n",
       "   'tnf',\n",
       "   'dis_tumor_dis',\n",
       "   'cell',\n",
       "   'type'))]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doc_id.take(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('10022814', 1), ('growth', 5731), ('inhibition', 2070)]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# counting by token\n",
    "# ('token', 1) --> ('token', [1++])\n",
    "reducer_2 = mapper_3.reduceByKey(lambda x, y: x+y)\n",
    "reducer_2.take(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert into python list so we can pass it into a map transformation\n",
    "doc_id1 = doc_id.map(list).collect()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('10022814', 9.277438115508676),\n",
       " ('growth', 0.6237928009569395),\n",
       " ('inhibition', 1.6421342292492622)]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idf = reducer_2.map(lambda x: (x[0], math.log(len(doc_id1)/x[1])))\n",
    "idf.take(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "list"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joint_my = shuffle.join(idf).take(5)\n",
    "joint_my\n",
    "type(joint_my)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parallelize so we can convert list back into a RDD\n",
    "rdd = sc.parallelize(joint_my)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, ('10022814', 1, 9.277438115508676, 9.277438115508676)),\n",
       " (0, ('growth', 6, 0.6237928009569395, 3.742756805741637)),\n",
       " (1, ('growth', 3, 0.6237928009569395, 1.8713784028708185)),\n",
       " (5, ('growth', 8, 0.6237928009569395, 4.990342407655516))]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ('doc_id', 'everything_else') --> (('doc_id'), ('token', 'tf', 'idf', 'tf-idf'))\n",
    "mapper_4 = rdd.map(lambda x: (x[1][0][0], ( x[0], x[1][0][1], x[1][1], \\\n",
    "    x[1][0][1] * x[1][1])) ).sortByKey()\n",
    "mapper_4.take(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.8 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e04e036266f305a312129840b06cfb353c1d2d005428a9f350cd8fe1c5454b5e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
